`load_json` and array values with `prefetch`

Hi Dan, would like the fix you merged here if possible https://github.com/piccolo-orm/piccolo/pull/1392

Would it be possible to release soon? If not, roughly when do you think you could?

Thank you

Thank you so much for your quick response Dan - however now I'm getting an issue with `json.loads` when the value is a list:

```
        if isinstance(s, str):
            if s.startswith('\ufeff'):
                raise JSONDecodeError("Unexpected UTF-8 BOM (decode using utf-8-sig)",
                                      s, 0)
        else:
            if not isinstance(s, (bytes, bytearray)):
>               raise TypeError(f'the JSON object must be str, bytes or bytearray, '
                                f'not {s.__class__.__name__}')
E               TypeError: the JSON object must be str, bytes or bytearray, not list
```


So on line 72 of `utils/objects.py` instead of:
```        
        elif load_json and key in json_column_names and value is not None:
```
May I suggest the following tweak:
```
        elif (
            load_json
            and key in json_column_names
            and (
                isinstance(value, str)
                or isinstance(value, bytes)
                or isinstance(value, bytearray)
            )
        ):
```

Thank you

--- a/piccolo/query/base.py
+++ b/piccolo/query/base.py
@@ -95,7 +95,7 @@ class Query(Generic[TableInstance, QueryResponseType]):
                 new_row = {**row}
                 for json_column_name in json_column_names:
                     value = new_row.get(json_column_name)
-                    if isinstance(value, (str, bytes, bytearray)):
+                    if value is not None:
                         new_row[json_column_name] = load_json(value)
                 processed_raw.append(new_row)
 

--- a/piccolo/query/base.py
+++ b/piccolo/query/base.py
@@ -95,7 +95,7 @@ class Query(Generic[TableInstance, QueryResponseType]):
                 new_row = {**row}
                 for json_column_name in json_column_names:
                     value = new_row.get(json_column_name)
-                    if value is not None:
+                    if isinstance(value, (str, bytes, bytearray)):
                         new_row[json_column_name] = load_json(value)
                 processed_raw.append(new_row)
 
@@ -105,28 +105,27 @@ class Query(Generic[TableInstance, QueryResponseType]):
 
         raw = await self.response_handler(raw)
 
-        if output:
-            if output._output.as_objects:
-                if output._output.nested:
-                    return cast(
-                        QueryResponseType,
-                        [
-                            make_nested_object(
-                                row,
-                                self.table,
-                                load_json=output._output.load_json,
-                            )
-                            for row in raw
-                        ],
-                    )
-                else:
-                    return cast(
-                        QueryResponseType,
-                        [
-                            self.table(**columns, _exists_in_db=True)
-                            for columns in raw
-                        ],
-                    )
+        if output and output._output.as_objects:
+            if output._output.nested:
+                return cast(
+                    QueryResponseType,
+                    [
+                        make_nested_object(
+                            row,
+                            self.table,
+                            load_json=output._output.load_json,
+                        )
+                        for row in raw
+                    ],
+                )
+            else:
+                return cast(
+                    QueryResponseType,
+                    [
+                        self.table(**columns, _exists_in_db=True)
+                        for columns in raw
+                    ],
+                )
 
         return cast(QueryResponseType, raw)
 

--- a/piccolo/utils/encoding.py
+++ b/piccolo/utils/encoding.py
@@ -63,7 +63,7 @@ class JSONDict(dict):
     ...
 
 
-def load_json(data: str) -> Any:
+def load_json(data: str | bytes | bytearray) -> Any:
     response = (
         orjson.loads(data) if ORJSON else json.loads(data)  # type: ignore
     )

--- a/piccolo/utils/objects.py
+++ b/piccolo/utils/objects.py
@@ -72,6 +72,10 @@ def make_nested_object(
         elif (
             load_json
             and key in json_column_names
+            # The top level might already have the JSON loaded, so check for
+            # a string, rather than it being non-null.
+            # https://github.com/piccolo-orm/piccolo/issues/1399
+            # Ideally we want a cleaner solution than this.
             and isinstance(value, (str, bytes, bytearray))
         ):
             table_params[key] = encoding.load_json(value)

--- a/piccolo/utils/objects.py
+++ b/piccolo/utils/objects.py
@@ -69,7 +69,11 @@ def make_nested_object(
             else:
                 # The value doesn't belong to a foreign key, so just append it.
                 table_params[key] = value
-        elif load_json and key in json_column_names and value is not None:
+        elif (
+            load_json
+            and key in json_column_names
+            and isinstance(value, (str, bytes, bytearray))
+        ):
             table_params[key] = encoding.load_json(value)
         else:
             table_params[key] = value

--- a/tests/table/test_join.py
+++ b/tests/table/test_join.py
@@ -459,6 +459,33 @@ class TestJoin(TableTest):
             instrument.recording_studio.facilities,
         )
 
+    def test_objects_nested_with_load_json_array(self):
+        """
+        Make sure that nested objects works alongside ``load_json``, when
+        the nested object has an array value for a JSON column.
+
+        https://github.com/piccolo-orm/piccolo/issues/1399#issuecomment-4864444701
+
+        """  # noqa: E501
+        RecordingStudio.update(
+            {
+                RecordingStudio.facilities: ["restaurant"],
+            },
+            force=True,
+        ).run_sync()
+
+        instrument = (
+            Instrument.objects(Instrument.recording_studio)
+            .output(load_json=True)
+            .first()
+            .run_sync()
+        )
+        assert instrument is not None
+        self.assertListEqual(
+            instrument.recording_studio.facilities,
+            ["restaurant"],
+        )
+
     def test_objects_prefetch_clause(self):
         """
         Make sure that ``prefetch`` clause works correctly.

--- a/tests/table/test_join.py
+++ b/tests/table/test_join.py
@@ -589,7 +589,7 @@ class TableB(Table):
 class TestObjectsNestedLoadJSONArray(AsyncTableTest):
     tables = [TableA, TableB]
 
-    async def test_objects_nested_with_load_json_parent(self):
+    async def test_objects_nested_with_load_json_array(self):
         """
         Make sure that nested objects works alongside ``load_json``, when
         the parent object has a JSON column, and it contains an array value.
@@ -604,11 +604,18 @@ class TestObjectsNestedLoadJSONArray(AsyncTableTest):
         table_a = TableA()
         await table_a.save()
 
-        table_b = TableB(data=["a", "b", "c"], table_a=table_a.id)
+        data = ["a", "b", "c"]
+        table_b = TableB(data=data, table_a=table_a.id)
         await table_b.save()
 
         # Without prefetch:
-        await TableB.objects().output(load_json=True)
+        row = await TableB.objects().output(load_json=True).first()
+        assert row
+        assert row.data == data
 
         # With prefetch:
-        await TableB.objects(TableB.table_a).output(load_json=True)
+        row = (
+            await TableB.objects(TableB.table_a).output(load_json=True).first()
+        )
+        assert row
+        assert row.data == data

--- a/tests/table/test_join.py
+++ b/tests/table/test_join.py
@@ -459,6 +459,7 @@ class TestJoin(TableTest):
             instrument.recording_studio.facilities,
         )
 
+    # TODO - move to a separate class
     def test_objects_nested_with_load_json_array(self):
         """
         Make sure that nested objects works alongside ``load_json``, when
@@ -469,7 +470,7 @@ class TestJoin(TableTest):
         """  # noqa: E501
         RecordingStudio.update(
             {
-                RecordingStudio.facilities: ["restaurant"],
+                RecordingStudio.facilities: ["restaurant", "toilets"],
             },
             force=True,
         ).run_sync()
@@ -483,7 +484,7 @@ class TestJoin(TableTest):
         assert instrument is not None
         self.assertListEqual(
             instrument.recording_studio.facilities,
-            ["restaurant"],
+            ["restaurant", "toilets"],
         )
 
     def test_objects_prefetch_clause(self):

--- a/tests/table/test_join.py
+++ b/tests/table/test_join.py
@@ -1,6 +1,9 @@
 import decimal
+from unittest import TestCase
 
-from piccolo.testing.test_case import TableTest
+from piccolo.columns.column_types import JSON, ForeignKey, Serial, Varchar
+from piccolo.table import Table
+from piccolo.testing.test_case import AsyncTableTest, TableTest
 from tests.base import engine_is
 from tests.example_apps.music.tables import (
     Band,
@@ -18,6 +21,9 @@ TABLES = [Manager, Band, Venue, Concert]
 
 class TestCreateJoin:
     def test_create_join(self):
+        """
+        A simple test to make sure tables with foreign keys can be created.
+        """
         for table in TABLES:
             table.create_table().run_sync()
 
@@ -74,9 +80,10 @@ class TestJoin(TableTest):
         )
         instrument.save().run_sync()
 
-    ###########################################################################
 
-    def test_join(self):
+class TestSelectJoin(TestJoin):
+
+    def test_select_join(self):
         select_query = Concert.select(
             Concert.band_1.name,
             Concert.band_2.name,
@@ -211,25 +218,6 @@ class TestJoin(TableTest):
                 },
             )
 
-    def test_proxy_columns(self):
-        """
-        Make sure that ``proxy_columns`` are set correctly.
-
-        There used to be a bug which meant queries got slower over time:
-
-        https://github.com/piccolo-orm/piccolo/issues/691
-
-        """
-        # We call it multiple times to make sure it doesn't change with time.
-        for _ in range(2):
-            self.assertEqual(
-                len(Concert.band_1._.manager._foreign_key_meta.proxy_columns),
-                2,
-            )
-            self.assertEqual(
-                len(Concert.band_1._foreign_key_meta.proxy_columns), 4
-            )
-
     def test_select_all_columns_root(self):
         """
         Make sure that using ``all_columns`` at the root doesn't interfere
@@ -343,7 +331,29 @@ class TestJoin(TableTest):
                 },
             )
 
-    ###########################################################################
+
+class TestProxyColumns(TestCase):
+    def test_proxy_columns(self):
+        """
+        Make sure that ``proxy_columns`` are set correctly.
+
+        There used to be a bug which meant queries got slower over time:
+
+        https://github.com/piccolo-orm/piccolo/issues/691
+
+        """
+        # We call it multiple times to make sure it doesn't change with time.
+        for _ in range(2):
+            self.assertEqual(
+                len(Concert.band_1._.manager._foreign_key_meta.proxy_columns),
+                2,
+            )
+            self.assertEqual(
+                len(Concert.band_1._foreign_key_meta.proxy_columns), 4
+            )
+
+
+class TestObjectsJoin(TestJoin):
 
     def test_objects_nested(self):
         """
@@ -412,80 +422,6 @@ class TestJoin(TableTest):
         self.assertIsInstance(ticket.concert.venue, Venue)
         self.assertIsInstance(ticket.concert.band_1.manager, Manager)
         self.assertIsInstance(ticket.concert.band_2.manager, Manager)
-
-    def test_objects_nested_with_load_json(self):
-        """
-        Make sure that nested objects works alongside ``load_json`` (i.e. the
-        JSON on nested objects gets loaded).
-
-        https://github.com/piccolo-orm/piccolo/issues/1383
-
-        """
-        instrument = (
-            Instrument.objects(Instrument.recording_studio)
-            .output(load_json=True)
-            .first()
-            .run_sync()
-        )
-        assert instrument is not None
-        self.assertDictEqual(
-            instrument.recording_studio.facilities,
-            {"restaurant": True},
-        )
-
-    def test_objects_nested_with_load_json_null(self):
-        """
-        Make sure that nested objects works alongside ``load_json``, when
-        the nested object has a null value for a JSON column.
-
-        https://github.com/piccolo-orm/piccolo/issues/1391
-
-        """
-        RecordingStudio.update(
-            {
-                RecordingStudio.facilities: None,
-            },
-            force=True,
-        ).run_sync()
-
-        instrument = (
-            Instrument.objects(Instrument.recording_studio)
-            .output(load_json=True)
-            .first()
-            .run_sync()
-        )
-        assert instrument is not None
-        self.assertIsNone(
-            instrument.recording_studio.facilities,
-        )
-
-    # TODO - move to a separate class
-    def test_objects_nested_with_load_json_array(self):
-        """
-        Make sure that nested objects works alongside ``load_json``, when
-        the nested object has an array value for a JSON column.
-
-        https://github.com/piccolo-orm/piccolo/issues/1399#issuecomment-4864444701
-
-        """  # noqa: E501
-        RecordingStudio.update(
-            {
-                RecordingStudio.facilities: ["restaurant", "toilets"],
-            },
-            force=True,
-        ).run_sync()
-
-        instrument = (
-            Instrument.objects(Instrument.recording_studio)
-            .output(load_json=True)
-            .first()
-            .run_sync()
-        )
-        assert instrument is not None
-        self.assertListEqual(
-            instrument.recording_studio.facilities,
-            ["restaurant", "toilets"],
-        )
 
     def test_objects_prefetch_clause(self):
         """
@@ -590,3 +526,81 @@ class TestJoin(TableTest):
         signing = Signing.objects().prefetch(Signing.with_).first().run_sync()
         assert signing is not None
         self.assertIsInstance(signing.with_, Band)
+
+
+class TestObjectsNestedLoadJSON(TestJoin):
+    def test_objects_nested_with_load_json(self):
+        """
+        Make sure that nested objects works alongside ``load_json`` (i.e. the
+        JSON on nested objects gets loaded).
+
+        https://github.com/piccolo-orm/piccolo/issues/1383
+
+        """
+        instrument = (
+            Instrument.objects(Instrument.recording_studio)
+            .output(load_json=True)
+            .first()
+            .run_sync()
+        )
+        assert instrument is not None
+        self.assertDictEqual(
+            instrument.recording_studio.facilities,
+            {"restaurant": True},
+        )
+
+    def test_objects_nested_with_load_json_null(self):
+        """
+        Make sure that nested objects works alongside ``load_json``, when
+        the nested object has a null value for a JSON column.
+
+        https://github.com/piccolo-orm/piccolo/issues/1391
+
+        """
+        RecordingStudio.update(
+            {
+                RecordingStudio.facilities: None,
+            },
+            force=True,
+        ).run_sync()
+
+        instrument = (
+            Instrument.objects(Instrument.recording_studio)
+            .output(load_json=True)
+            .first()
+            .run_sync()
+        )
+        assert instrument is not None
+        self.assertIsNone(
+            instrument.recording_studio.facilities,
+        )
+
+
+class TableA(Table):
+    id = Serial(primary_key=True)
+    name = Varchar()
+
+
+class TableB(Table):
+    data = JSON()
+    table_a = ForeignKey(TableA, null=False)
+
+
+class TestObjectsNestedLoadJSONParent(AsyncTableTest):
+    tables = [TableA, TableB]
+
+    async def test_objects_nested_with_load_json_parent(self):
+        """
+        Make sure that nested objects works alongside ``load_json``, when
+        the parent object has a JSON column.
+
+        https://github.com/piccolo-orm/piccolo/issues/1399#issuecomment-4864444701
+
+        """  # noqa: E501
+        table_a = TableA()
+        await table_a.save()
+
+        table_b = TableB(data={"a": 1}, table_a=table_a.id)
+        await table_b.save()
+
+        await TableB.objects(TableB.table_a).output(load_json=True)

--- a/tests/table/test_join.py
+++ b/tests/table/test_join.py
@@ -586,21 +586,29 @@ class TableB(Table):
     table_a = ForeignKey(TableA, null=False)
 
 
-class TestObjectsNestedLoadJSONParent(AsyncTableTest):
+class TestObjectsNestedLoadJSONArray(AsyncTableTest):
     tables = [TableA, TableB]
 
     async def test_objects_nested_with_load_json_parent(self):
         """
         Make sure that nested objects works alongside ``load_json``, when
-        the parent object has a JSON column.
+        the parent object has a JSON column, and it contains an array value.
 
         https://github.com/piccolo-orm/piccolo/issues/1399#issuecomment-4864444701
+
+        The problem was because we already loaded the JSON string on the top
+        level row, and then were trying to load it again in
+        `make_nested_object`.
 
         """  # noqa: E501
         table_a = TableA()
         await table_a.save()
 
-        table_b = TableB(data={"a": 1}, table_a=table_a.id)
+        table_b = TableB(data=["a", "b", "c"], table_a=table_a.id)
         await table_b.save()
 
+        # Without prefetch:
+        await TableB.objects().output(load_json=True)
+
+        # With prefetch:
         await TableB.objects(TableB.table_a).output(load_json=True)

